### ISSUE TITLE
agent splay sync requests to server

### ISSUE THREAD

**Issue by amoore877** [spiffe/spire#1240]
To avoid a stampede of agents contacting the server on startup or recovery, agents should splay their syncing a bit with jitter.

Similar logic could piggy back on this and implement backoff.

**Comment by azdagron** [comment]
In general, jittering intervals is a good idea. However, considering agents sync roughly every five seconds, how much jitter here would even be helpful/noticeable? The variability in agent startup times (probably on the order of seconds) would probably get pretty close in introducing natural jitter considering the 5 second interval...

**Comment by APTy** [comment]
There is a subtlety here I should probably address. There are two intervals:

1. Agent Poll: the agent polls the server for any updated CSRs or stale local SVIDs (every 5 seconds)
2. Agent CSR: the agent sends the server a list of CSRs to be signed (every 1/2 SVID TTL)

Jitter on (1) would be fine, but as you say, some natural jitter may exist anyway.

Jitter on (2) is more important. Imagine you have a registrant that causes 100 registrations to be created on the Server at time t=0 with SVID TTL=60m. If you have 1,000 Agents in your cluster, this will result in about 100,000 signing operations in a ~5 second period at time t=0, then again at t=30m, and so on.

This type of bursty load is _maybe_ acceptable when a cluster first comes online; however, it's not preferable in the steady state. If, for example, SVIDs were considered "stale" on a random interval between (0 * SVID TTL, 0.5 * SVID TTL), it would smooth out the interval of signing operations, allowing load on the server to be more evenly distributed across time.

**Comment by amoore877** [comment]
@azdagron 
Replying to https://github.com/spiffe/spire/pull/1252#issuecomment-554560411

Requests to renew a SVID are directly based on the SVIDs lifetime / remaining TTL. Common functionality for checking this (`are we at half TTL or less?`) was in multiple places, one happening to use `<` and the other `<=`. Centralizing:
- Benefit of reduced code, and lessens chances of logical discrepancies
- Can be used to have SVID rotation jitter at the `shouldRotate` logic level. Today this is the logic of `half TTL`  implies `shouldRotate`, future work can change this.

Just throwing all the thoughts out there:

**Idea A:** (I believe in line with @APTy's suggestion)
Change to `Jittered% TTL` as a limit.
The jitter should have a hard bound of 50% on one end (always return `true` for `shouldRotate` if TTL is 50% or less), and we'll have to think on a reasonable opposing bound. This *likely* make sense as a configurable.

Implications of this would be:
- statistically, renewals would occur earlier; agent and workload SVIDs will on average have more lifetime
- on a failed update, the same SVID may not be jittered into renewal again (unless at half TTL), which is a bit inconsistent
- SVID TTL, jitter range, and sync / rotate intervals have a critical interaction-
-- Very frequent sync intervals mean that statistically most SVIDs will be renewed at the upper bound of the jitter range (since the jitter would happen on every call of `shouldRotate`), probably reducing splay.
-- (as is true today) Sync intervals that are too close to scale of TTL levels (eg., 10m and 60m) means we probably expect reduce splaying of renewals
-- A large amount of jitter means requests to Server will come in much more frequently, increasing overall load, though with reasonable intervals this will at least still be splayed well.
-- Very low jitter range negates the new logic completely, unless SVID TTL is very large.

**Idea B:** (also in line with @APTy's suggestion)
On fetching, every SVID on agent end gets an additional jittered configuration for at what point in TTL they should actually start renewing (latest of 50%, as today). `shouldRotate` logic checks this field instead of / in addition to 50% check. This is similar to **A** except the rotation time is consistent for each SVID on an agent, rather than per-run of `shouldRotate`
- makes agent logic more complex with an in-memory concurrent mapping of SVID-to-TTL-%
- better avoids complex interplay of the new functionality with TTL and sync intervals

**Idea C:**
Change to `half TTL OR %earlyRenewalChance`.
Some form of early renewal lottery. From my perspective I rejected this since the lottery chance must be infinitesimally small for the default sync interval of 5 seconds, which becomes hard to reason about.

**Idea D:**
(unrelated to work in #1252 )
Some form of sometimes giving a reduced TTL SVID. From my perspective I rejected this since it violates expected behavior from configuration and puts in environment in a less known/reliable state.
